New LUKS devices management module

japokorn opened the pull request:
New module that uses `cryptsetup` (LUKS) functions for management of encrypted devices.
Supports creation/removing/locking/unlocking of LUKS containers as well as adding and removing of additional keys. Unit tests included.
This module is needed for storage operations with encrypted devices.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
luks_device

##### ADDITIONAL INFORMATION
Ensure that your keyfiles are protected.

japokorn committed:
New LUKS devices management module

- new module that uses cryptsetup (LUKS) functions for management
of encrypted devices
- unit tests included

felixfontein commented:
@japokorn thanks for your PR!

Could you please fix the issues with the sanity checks?

Some general comments:
 - Why not name this module `luks_device` instead of `crypt_device`? `crypt_device` sounds very generic.
 - You recommend to use Ansible Vault, but require the keys to be written to disk. It would be useful to be able to provide the keys directly via the playbook (encrypted by Vault or not), so that they are directly piped into `cryptsetup` without being written to disk (`cryptsetup` supports that).

felixfontein commented on lib/ansible/modules/crypto/crypt_device.py:
```diff
@@ -0,0 +1,444 @@
+    device:
+        description:
+            - "Device to work with (e.g. '/dev/sda1'). Needed in most cases."
+        type: string
```
You need to use `str`, not `string`.

felixfontein commented on lib/ansible/modules/crypto/crypt_device.py:
```diff
@@ -0,0 +1,444 @@
+
+short_description: Manage encrypted (LUKS) devices
+
+version_added: "2.5"
```
The next release of Ansible will be 2.8.

felixfontein commented on lib/ansible/modules/crypto/crypt_device.py:
```diff
@@ -0,0 +1,444 @@
+    # available arguments/parameters that a user can pass
+    module_args = dict(
+        device=dict(type='str', required=False),
+        state=dict(type='str', required=False, default='present'),
```
You need to add `choices=['absent', 'present']` to the `state` `dict`.

felixfontein commented on lib/ansible/modules/crypto/crypt_device.py:
```diff
@@ -0,0 +1,444 @@
+        super(Conditions, self).__init__(module)
+
+    def luks_create(self):
+        return self._module.params['device'] is not None and\
```
If you use round brakets, you don't need to use `\`.

japokorn commented:
@felixfontein Thanks for the quick reaction. I didn't expect to someone actually look at this so fast. :)
I have fixed most of the reported issues except two:

1. `E326 Value for "choices" from the argument_spec ([]) for "state" does not match the documentation (['present', 'absent'])` This I have the same way the other modules do. (Am I missing something or is it a bug?)

2. `E121 continuation line under-indented for hanging indent` It is not possible to indent these ifs so PEP8 doesn't complain. No way described in PEP8 documentation works nor any other one I could come up with.

Please check your CI settings or let me know what to do to make it work. Thanks.

felixfontein commented:
>     1. `E326 Value for "choices" from the argument_spec ([]) for "state" does not match the documentation (['present', 'absent'])` This I have the same way the other modules do. (Am I missing something or is it a bug?)

Which modules do you mean? Every module I checked uses `choices` in `argument_spec` if it is used in the documentation.

>     2. `E121 continuation line under-indented for hanging indent` It is not possible to indent these ifs so PEP8 doesn't complain. No way described in PEP8 documentation works nor any other one I could come up with.

Can you try this?
``` .py
         if (self._module.params['device'] is None or
                self._module.params['key'] is None or
                self._module.params['open'] is False):
            # conditions for open not fulfilled
            return False 
```

BTW, if you install flake8 on your machine (`pip install flake8`), you can run `flake8 /path/to/your/module.py` and get (almost) instant feedback on PEP8 issues.

> Please check your CI settings or let me know what to do to make it work. Thanks.

I don't have any CI settings. Do you mean the CI used by Ansible?

japokorn commented:
@felixfontein Ok, I have misinterpreted the first issue and the indentation seems to be working now. Curious. Yesterday it didn't work when I tried it this way. Never mind. Thanks for your help.

felixfontein commented on lib/ansible/modules/crypto/crypt_device.py:
```diff
@@ -0,0 +1,446 @@
+        device=dict(type='str', required=False),
+        open=dict(type='bool', required=False, default=False),
+        name=dict(type='str', required=False),
+        key=dict(type='str', required=False),
```
For paths, you can use `type='path'`. Then Ansible will for example resolve `~` in the path if the user uses that.

japokorn replied:
Fixed.

felixfontein commented on lib/ansible/modules/crypto/crypt_device.py:
```diff
@@ -0,0 +1,446 @@
+
+DOCUMENTATION = '''
+---
+module: crypt_device
```
What do you think of changing the name of this module to `luks_device`? That would be more specific, and makes it easier to find and remember. Also, the name `crypt_device` is very generic and does not hint that this module is Linux or even LUKS specific.

japokorn replied:
Name changed.

felixfontein commented on lib/ansible/modules/crypto/crypt_device.py:
```diff
@@ -0,0 +1,446 @@
+            - "Used to unlock the container and needed for most of the operations.
+              Parameter value is the path to keyfile with the passphrase.
+              BEWARE that working with keyfiles in plaintext is dangerous.
+              Use Ansible Vault to protect them."
```
Ansible vault does not help you with files written to disk on the target machine. It only allows you to store them encrypted on the host. If you want to encourage Vault usage, you should provide an option which allows to provide `key` as text. I would add two options: `key_path` and `key_content` (mutually exclusive). The first would be a path to the key on disk, the second would allow to directly pass the passphrase into the module (from a Vault file, for example).

felixfontein replied:
Removing the Vault part is also OK; and the `key_path` / `key_content` thing can also be implemented in another PR.

japokorn replied:
After some thinking I have decided to not mention Vault at all. At the moment I am not sure about security of usage `key_content`.

japokorn replied:
Adding `key_content` would also require adding `new_key_content` and `remove_key_content` options to stay coherent. That would lead to much more complex evaluation logic. I do agree with you that it would help to have this option but I would rather make it a future enhancement.

japokorn replied:
After giving it some thought: It should be possible to check whether supplied string is a path to some existing file or not. Based on this check the `key` option can accept both: passphrase and path to a keyfile and then resolve it automatically. Still, I would rather add it in a future when this version proves its worth.

felixfontein replied:
Yes, it would also need `new_key_content` and `remove_key_content` at some point. And yes, let's do stuff like this in the future :)

About the idea of `key` doing some kind of detection magic: I don't think this is a good idea. Someone might actually use pathnames (or something which looks like a pathname) as a passphrase (maybe because it's easier to memorize and fulfils some "must contain other character" condition?). From experience, sooner or later this always becomes a problem, and once we are at that point, it's really hard to fix it without breaking backwards compatibility.

I think just adding `_path` to all these options is probably the best thing right now, even just because it makes it less likely that a user accidentally puts the key content itself in there (and then it's in the logs, since paths usually aren't marked as `no_log`).

japokorn replied:
Danger of getting passphrases into logs by mistake is a strong argument. I have changed option names to `keyfile`, `new_keyfile` and `remove_keyfile`.

felixfontein commented on lib/ansible/modules/crypto/crypt_device.py:
```diff
@@ -0,0 +1,446 @@
+options:
+    device:
+        description:
+            - "Device to work with (e.g. '/dev/sda1'). Needed in most cases."
```
You can use `C(/dev/sda1)` for formatting.

felixfontein replied:
You should also be more precise in *which* cases `device` (and the other options) are needed.

japokorn replied:
Fixed.

felixfontein commented on lib/ansible/modules/crypto/crypt_device.py:
```diff
@@ -0,0 +1,446 @@
+        type: bool
+    name:
+        description:
+            - "Sets container name when 'open' is true. Can be used
```
`Sets container name when I(open) is C(true).` (See https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#linking-within-module-documentation for some more formatting directives.)

japokorn replied:
Fixed.

felixfontein commented on lib/ansible/modules/crypto/crypt_device.py:
```diff
@@ -0,0 +1,446 @@
+
+- name: close the LUKS container "mycrypt"
+  crypt_device:
+    open: "false"
```
Booleans do not require quotation marks in YAML. In fact, this defines a string with value `false`. It is only because Ansible does some magic that this is converted back to a boolean.

japokorn replied:
Fixed.

felixfontein commented on lib/ansible/modules/crypto/crypt_device.py:
```diff
@@ -0,0 +1,446 @@
+        type: str
+
+requirements:
+    - "cryptsetup"
```
You should also mention `wipefs`.

felixfontein replied:
And `lsblk`.

japokorn replied:
Fixed.

felixfontein commented on lib/ansible/modules/crypto/crypt_device.py:
```diff
@@ -0,0 +1,446 @@
+
+    def run_luks_create(self, device, keyfile):
+        # create a new luks container; use batch mode to auto confirm
+        result = self._run_command(['cryptsetup', 'luksFormat',
```
Instead of assuming that `cryptsetup` (and below, `wipefs`) can be called without specifying a path, you should use `self._module.get_bin_path('cryptsetup', True)` (the `True` makes Ansible die if it cannot find the executable). It makes sense to do it once in `__init__()` or `cryptsetup` (for `wipefs`, I guess you can do it where you need it).

japokorn replied:
Fixed.

felixfontein commented on lib/ansible/modules/crypto/crypt_device.py:
```diff
@@ -0,0 +1,446 @@
+    crypt = CryptHandler(module)
+    conditions = Conditions(module)
+
+    if module.params['state'] not in [None, 'present', 'absent']:
```
This is not needed. Ansible will always make `module.params['state']` be either `'present'` or `'absent'`. If the user tries to specify something else, `AnsibleModule()` will fail.

In general, you should avoid doing parameter verification manually as much as possible, and use Ansible's facilities for that. Check out the other arugments of `AnsibleModule.__init__()` and how they are used in other modules.

japokorn replied:
Fixed.

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -0,0 +1,448 @@
+            - "Used to unlock the container and needed for most
+              of the operations. Parameter value is the path
+              to the keyfile with the passphrase.
+              BEWARE that working with keyfiles in plaintext is dangerous.
```
Maybe you should put the `BEWARE` part into a new paragraph (i.e. a new list item in `description`).

japokorn replied:
Fixed.

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -0,0 +1,448 @@
+version_added: "2.8"
+
+description:
+    - "Module manages LUKS on given device. Supports creating, destroying,
```
Maybe make the first LUKS a link to https://en.wikipedia.org/wiki/Linux_Unified_Key_Setup:
```suggestion
    - "Module manages L(LUKS,https://en.wikipedia.org/wiki/Linux_Unified_Key_Setup) on given device. Supports creating, destroying,
```

japokorn replied:
Fixed.

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -0,0 +1,448 @@
+RETURN = '''
+name:
+    description:
+        When 'open' option is used returns (generated or given) name
```
You can also use `I(...)` here.

japokorn replied:
Fixed.

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -0,0 +1,448 @@
+        type: path
+
+requirements:
+    - "cryptsetup, wipefs, lsblk"
```
It's probably better to put each requirement into its own item.

japokorn replied:
Fixed.

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -0,0 +1,448 @@
+
+    def __init__(self, module):
+        self._module = module
+        self._module.get_bin_path('cryptsetup', True)
```
You should store the result in a variable (e.g. `self._cryptsetup_bin`).

japokorn replied:
Fixed.

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -0,0 +1,448 @@
+
+    def run_luks_create(self, device, keyfile):
+        # create a new luks container; use batch mode to auto confirm
+        result = self._run_command(['cryptsetup', 'luksFormat',
```
Don't write `cryptsetup`, instead use `self._cryptsetup_bin` (or whatever variable you stored it into above). Otherwise, calling `get_bin_path()` is pretty much superfluous.

japokorn replied:
Fixed.

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -0,0 +1,448 @@
+            raise ValueError('Error while closing LUKS container %s' % (name))
+
+    def run_luks_remove(self, device):
+        self._module.get_bin_path('wipefs', True)
```
You also need to store this into a variable, otherwise you can't use the resulting path.

japokorn replied:
Fixed.

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -0,0 +1,448 @@
+              instead of I(device) when closing the container
+              (i.e. when I(open=false))."
+        type: str
+    key:
```
What do you think about changing this option name from `key` to `key_path`? If it is just `key`, I would expect to be able to put the key itself in here, and not a path to the keyfile.

japokorn replied:
Since there are two similar options (`new_key` and `remove_key`) that would need the same approach to stay coherent, that would lead to misleading names (`new_key path` and `remove_key_path`). See your `key_path`/`key_content` comment for proposed solution.

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -0,0 +1,448 @@
+        return device
+
+
+class Conditions(CryptHandler):
```
Instead of deriving from `CryptHandler`, why not pass `CryptHandler` as a constructor argument?

Otherwise, I don't see why you need both a `CryptHandler` and a `Conditions` instance in the main module code (since you can use the `Conditions` instance for both).

japokorn replied:
Those two classes do different things and need to be distinguished. However, I agree that this solution is not optimal. To address this I did a bit of rework: Now there are three classes:
`Handler` - basic class that takes care of generic stuff both of the former classes needed.
`CryptHandler` - inherits from `Handler`, generic stuff was moved to it from here.
`ConditionsHandler` - inherits from `Handler`, what was formerly needed from old `CryptHandler` is now moved to `Handler`.

I think that this approach is much more elegant and simplifies understanding of the workings of the module.

felixfontein replied:
Yes, that's a better solution.

BTW, why not put all functionality calling `cryptsetup` into `CryptHandler`, and giving `ConditionsHandler` a reference to `CryptHandler` for the calls it needs to do? Then there's a more clear separation into `CryptHandler`, which does all the `cryptsetup` calls, and `ConditionsHandler`, which contains most of the decision logic.

japokorn replied:
Fixed.

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -0,0 +1,448 @@
+        try:
+            crypt.run_luks_create(module.params['device'],
+                                  module.params['key'])
+        except ValueError as e:
```
Why not have a large `try: except` block around all the operations which catches `ValueError`? That would improve readability alot.

japokorn replied:
It is a good practice to have `try: except` block covering only the critical part of the code. This minimizes risk of catching unwanted exceptions at the cost of slightly lower readibility.

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -0,0 +1,448 @@
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
+    ---
```
I think you need to remove the spaces before `---`; YAML doesn't seem to like that.

felixfontein replied:
This (and the whitespace error in line 43) are probably the only things the sanity tests complain about.

japokorn replied:
Fixed.

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -0,0 +1,457 @@
+
+    # luks close
+    if conditions.luks_close():
+        if module.params['device'] is not None:
```
Why do you have no `try/except` block around these calls?

japokorn replied:
`conditions.luks_close` does not raise any exception to be caught.

felixfontein replied:
The line below calls `get_container_name_by_device`, which can raise `ValueError` according to its documentation.

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -0,0 +1,457 @@
+
+    # luks remove
+    if conditions.luks_remove():
+        crypt.run_luks_remove(module.params['device'])
```
There's another `try/except` missing here.

japokorn replied:
`conditions.luks_remove` does not raise any exception to be caught.

felixfontein replied:
`run_luks_remove()` does raise `ValueError`.

felixfontein replied:
Here's another line which can raise `ValueError` (call to `run_luks_remove()`).

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -0,0 +1,460 @@
+
+    def __init__(self, module):
+        self._module = module
+        self._lsblk_bin = self._module.get_bin_path('lsblk', true)
```
```suggestion
        self._lsblk_bin = self._module.get_bin_path('lsblk', True)
```

felixfontein commented:
One general remark: instead of having `state` (`present`/`absent`) and `open` (`true`/`false`), why not have four values for `state`? I.e. `present`, `opened`, `closed`, and `absent`? `present` simply makes sure it is there (and doesn't care about being open or closed), `opened` and `closed` imply `present`, and `absent` makes sure it is closed and wiped? This would simplify logic a bit (and avoid errors such as `Contradiction in setup: LUKS set to be 'absent' and 'open'.`).

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -0,0 +1,460 @@
+        name = self._crypthandler.get_container_name_by_device(
+            self._module.params['device'])
+        if name is not None:
+            if name != self._module.params['name']:
```
If `self._module.params['name'] is None`, this will never match and the module fails.

japokorn replied:
Fixed.

felixfontein commented:
Looks good besides the problem described above; I've used the following playbook to test it a bit:
```.yaml
---
- hosts: localhost
  tasks:
  - command: dd if=/dev/zero of=cryptfile bs=1M count=32
  - command: losetup -f cryptfile
    become: yes
  - shell: losetup -a | grep cryptfile | sed -e 's/:.*//g'
    become: yes
    register: cryptfile_device_output
  - set_fact:
      cryptfile_device: "{{ cryptfile_device_output.stdout }}"
  - copy:
      content: "asdf"
      dest: keyfile
  - name: Create
    luks_device:
      device: "{{ cryptfile_device }}"
      state: present
      open: false
      keyfile: keyfile
    become: yes
  - name: Create (idempotent)
    luks_device:
      device: "{{ cryptfile_device }}"
      state: present
      open: false
      keyfile: keyfile
    become: yes
  - name: Open
    luks_device:
      device: "{{ cryptfile_device }}"
      state: present
      open: true
      keyfile: keyfile
    become: yes
    register: luks_open
  - name: Open (idempotent)
    luks_device:
      device: "{{ cryptfile_device }}"
      state: present
      open: true
      keyfile: keyfile
    become: yes
  - name: Closed (via name)
    luks_device:
      name: "{{ luks_open.name }}"
      open: false
    become: yes
  - name: Closed (via device, idempotent)
    luks_device:
      device: "{{ cryptfile_device }}"
      open: false
    become: yes
  - name: Opened
    luks_device:
      device: "{{ cryptfile_device }}"
      state: present
      open: true
      keyfile: keyfile
    become: yes
  - name: Absent
    luks_device:
      device: "{{ cryptfile_device }}"
      state: absent
    become: yes
  - name: Absent (idempotence)
    luks_device:
      device: "{{ cryptfile_device }}"
      state: absent
    become: yes
  - command: losetup -d "{{ cryptfile_device }}"
    become: yes
  - file:
      dest: cryptfile
      state: absent
```

felixfontein commented:
The module should really return the name of the opened container if it already was opened. The code to determine the name is already there and already executed, it's just that the result is not returned.

japokorn committed:
New LUKS devices management module

- modified interface by removing 'open' option and moving its functionality
into 'state' option

japokorn commented:
Implemented `state` and `open` change. `open` option was removed and its functionality moved to `state`. Supported values are now `present`, `absent`, `opened`, `closed`.
The module now also returns the name of the already opened container when `opened` option is used.

Since this required quite a change for both code and the tests, I have submitted it into another commit. This should allow easier comparison with the previous one. The commits are to be squashed together before the final merge.

felixfontein commented:
Cool! I'll probably manage to take a closer look over the weekend.

> Since this required quite a change for both code and the tests, I have submitted it into another commit. This should allow easier comparison with the previous one.

Nowadays (with the force-merge diffs github is showing) it's not that hard anymore. But an additional commit is nice anyway :)

> The commits are to be squashed together before the final merge.

The commits are always squashed when merged for Ansible (the only exception is if there are different authors, then in special situations it can be that each author gets one commit in the final result).

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -0,0 +1,512 @@
+              as a part of the result."
+            - "I(closed) will lock the LUKS container. However if the container
+              does not exist it will be created.
+              Requires I(device) and I(keyfile) options to be provided. If
```
Closing does not need `keyfile`. (Tested it with the module, works fine without `keyfile` option.)

japokorn replied:
That is technically true. However `closed` can now also create a LUKS container if it does not exist. And container creation requires `device` and `keyfile`. In case the user is absolutely sure that container is already present only `device` or even just `name` will do. Next sentence in the doc explains that.

felixfontein replied:
Ah, I misread that. Totally makes sense :)

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -0,0 +1,512 @@
+
+    # luks close
+    try:
+        if conditions.luks_close():
```
For consistency reasons, why not have the `try:` inside the `if ...:` block? (Or just have one global `try` for all operations ;-) )

japokorn replied:
Oops. Completely missed that one. Thank you for being so thorough. Fixed.

felixfontein replied:
You're welcome. I found two more, btw ;-)

felixfontein reviewed (commented):
Looks good (except two minor things, see comments). I tested it successfully with the following playbook (similar to above, but adjusted to changed options):
```.yaml
---
# ansible-playbook -K test-luks_device.yml 
- hosts: localhost
  tasks:
  - command: dd if=/dev/zero of=cryptfile bs=1M count=32
  - command: losetup -f cryptfile
    become: yes
  - shell: losetup -a | grep cryptfile | sed -e 's/:.*//g'
    become: yes
    register: cryptfile_device_output
  - set_fact:
      cryptfile_device: "{{ cryptfile_device_output.stdout }}"
  - block:
    - copy:
        content: "asdf"
        dest: keyfile
    - name: Create
      luks_device:
        device: "{{ cryptfile_device }}"
        state: present
        keyfile: keyfile
      become: yes
    - name: Create (idempotent)
      luks_device:
        device: "{{ cryptfile_device }}"
        state: present
        keyfile: keyfile
      become: yes
    - name: Open
      luks_device:
        device: "{{ cryptfile_device }}"
        state: opened
        keyfile: keyfile
      become: yes
      register: luks_open
    - name: Open (idempotent)
      luks_device:
        device: "{{ cryptfile_device }}"
        state: opened
        keyfile: keyfile
      become: yes
    - name: Closed (via name)
      luks_device:
        name: "{{ luks_open.name }}"
        state: closed
      become: yes
    - name: Closed (via device, idempotent)
      luks_device:
        device: "{{ cryptfile_device }}"
        state: closed
      become: yes
    - name: Opened
      luks_device:
        device: "{{ cryptfile_device }}"
        state: opened
        keyfile: keyfile
      become: yes
    - name: Absent
      luks_device:
        device: "{{ cryptfile_device }}"
        state: absent
      become: yes
    - name: Absent (idempotence)
      luks_device:
        device: "{{ cryptfile_device }}"
        state: absent
      become: yes

    always:
    - command: losetup -d "{{ cryptfile_device }}"
      become: yes
    - file:
        dest: cryptfile
        state: absent
```

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -0,0 +1,512 @@
+    # luks close
+    if conditions.luks_close():
+        if module.params['device'] is not None:
+            name = crypt.get_container_name_by_device(module.params['device'])
```
This line can also raise a `ValueError`.

japokorn replied:
Fixed.

japokorn replied:
Fixed.

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -0,0 +1,516 @@
+
+    # luks remove
+    if conditions.luks_remove():
+        crypt.run_luks_remove(module.params['device'])
```
Here's still another one.

japokorn replied:
Fixed.

felixfontein commented:
There's still a `try`/`except` missing in line 504:
https://github.com/ansible/ansible/blob/727cb0f464800725d9c6e6d1ebc1573efd13250b/lib/ansible/modules/crypto/luks_device.py#L502-L505
In case you really already fixed that, you forgot to push the commit.

japokorn commented:
I haven't seen your second comment. It should be ok now.

felixfontein reviewed (approved):
I think this is good now! I've tested it again with the same playbook as before, and it still seems to work ;-)

shipit

felixfontein commented:
@Spredzy @MarkusTeufelberger @thomwiggers @lolcube feel free to also review this ;-)

felixfontein commented:
I've added the PR to a list of PRs to be reviewed on the next Monthly Big PR review (will be held on February 21st). Maybe there will already be more reviews before that, but maybe not :)

felixfontein commented:
bot_status

(hopefully convinces ansibot to add the `crypto` label...)

MarkusTeufelberger commented:
Sorry, I'm not really an expert for LUKS...

Pinging @tux21b who might be interested in this module for his playbooks.

felixfontein commented:
@Xyon since you've indicated you're interested in LUKS, you might be interested in this PR as well!

Xyon commented:
> @Xyon since you've indicated you're interested in LUKS, you might be interested in this PR as well!

I am - and I can review, but I don't think I have bot access to approve yet.

Xyon commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -0,0 +1,519 @@
+              Parameter value is the path to the keyfile with the passphrase."
+            - "BEWARE that it is possible to remove even the last key from the
+              container. Data in there will be irreversibly lost
+              without a warning."
```
I don't know if I like this approach. I can't think of a time when this would ever be desired behaviour, and not a mistake - even so, I feel like we should require a force flag to remove the *only* key on a container, rendering it useless.

felixfontein replied:
Good question :) If you have a backup of the real key, you can still access the volume afterwards (though only by working around LUKS); even though this is a very exotic use-case, I wouldn't disallow this. Thus having a flag (with default value "prevent this") would definitely be helpful. Maybe something called "allow_to_remove_last_key" with default value `false`?

gundalow replied:
something like `allow_to_remove_last_key` (default false) sounds good.

Xyon replied:
Yep - I'm not against the _notion_ that it be possible, it just shouldn't be possible _accidentally_. Most LUKS containers I've worked with only have one key on them, ever.

felixfontein replied:
I can add that in a subsequent PR, so we don't have to delay merging because of this :)

Xyon replied:
Sounds fine to me. I don't think I can say shipit yet, but if this gets wrapped up in another PR the rest looks fine to me.

Xyon commented:
Conversation above considered, LGTM. 

shipit

gundalow commented:
Anyone else able to review this, otherwise we will merge soon. Thanks in advance!

puiterwijk commented:
With the follow-up PR to require a separate flag for removing the last key, LGTM.

felixfontein commented:
@japokorn thank you very much for this PR! The `luks_device` module will show up in Ansible 2.8, to be released [this spring](https://docs.ansible.com/ansible/devel/roadmap/ROADMAP_2_8.html#expected).
@Xyon @puiterwijk thanks for reviewing!
@gundalow thanks for merging!

felixfontein commented:
I created two follow-up PRs:
1. integration tests in #52359;
2. `allow_to_remove_last_key` option in #52371.
